test/integration/cert_storage_failed_test.go

```diff
@@ -0,0 +1,198 @@
+	// would fail to revoke the certificate later.
+	_, err = db.Exec(`
+		CREATE TRIGGER fail_ready
+		BEFORE UPDATE ON certificateStatus
+		FOR EACH ROW BEGIN
+		DECLARE reversedName1 VARCHAR(255);
+		SELECT reversedName
+		    INTO reversedName1
+			FROM issuedNames
+			WHERE serial = NEW.serial
+			    AND reversedName LIKE "com.wantserror.%";
+		IF NEW.status = "good" AND reversedName1 != "" THEN
+			SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Pretend there was an error updating the certificateStatus';
+		END IF;
+		END
+	`)
+	test.AssertNotError(t, err, "failed to create trigger")
+	
+	defer db.Exec(`DROP TRIGGER IF EXISTS fail_ready`)
+
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	test.AssertNotError(t, err, "creating random cert key")
+
+	// ---- Test revocation by serial ----
+	revokeMeDomain := "revokeme.wantserror.com"
+	// This should fail because the trigger prevented setting the certificate status to "ready"
+	_, err = authAndIssue(nil, certKey, []string{revokeMeDomain}, true)
+	test.AssertError(t, err, "expected authAndIssue to fail")
+
+	cert, err := getPrecertByName(db, revokeMeDomain)
+	test.AssertNotError(t, err, "failed to get certificate by name")
+
+	err = expectOCSP500(cert)
+	test.AssertNotError(t, err, "expected 500 error from OCSP")
+
+	// Revoke by invoking admin-revoker
+	config := fmt.Sprintf("%s/%s", os.Getenv("BOULDER_CONFIG_DIR"), "admin-revoker.json")
+	output, err := exec.Command("./bin/admin-revoker", "serial-revoke",
+		"-config", config,
+		core.SerialToString(cert.SerialNumber),
+		"0").CombinedOutput()
+	test.AssertNotError(t, err, fmt.Sprintf("revoking via admin-revoker: %s", string(output)))
+
+	_, err = ocsp_helper.Req(cert,
+			ocsp_helper.DefaultConfig.WithExpectStatus(ocsp.Revoked).WithExpectReason(ocsp.Unspecified))
+
+	// ---- Test revocation by key ----
+	blockMyKeyDomain := "blockmykey.wantserror.com"
+	// This should fail because the trigger prevented setting the certificate status to "ready"
+	_, err = authAndIssue(nil, certKey, []string{blockMyKeyDomain}, true)
+	test.AssertError(t, err, "expected authAndIssue to fail")
+
+	cert, err = getPrecertByName(db, blockMyKeyDomain)
+	test.AssertNotError(t, err, "failed to get certificate by name")
+
+	err = expectOCSP500(cert)
+	test.AssertNotError(t, err, "expected 500 error from OCSP")
+
+	// Time to revoke! We'll do it by creating a different, successful certificate
+	// with the same key, then revoking that certificate for keyCompromise.
+	revokeClient, err := makeClient()
+	test.AssertNotError(t, err, "creating second acme client")
+	res, err := authAndIssue(nil, certKey, []string{random_domain()}, true)
+	test.AssertNotError(t, err, "issuing second cert")
+
+	successfulCert := res.certs[0]
+	err = revokeClient.RevokeCertificate(
+		revokeClient.Account,
+		successfulCert,
+		certKey,
+		1,
+	)
+	test.AssertNotError(t, err, "revoking second certificate")
+	
+	for i := 0; i < 30; i++ {
+		time.Sleep(50 * time.Millisecond)
```

Comment:
Put the sleep after the first query, so it's _possible_ to complete this in less than 50ms if bad-key-revoker ran at the right time?

Also, optionally do this query much faster than bad-key-revoker's cycle time, so that we're not waiting a whole 49ms after it's done its work?